web_poet/example.py

```diff
@@ -0,0 +1,76 @@
+from asyncio import run
+from typing import Any, Dict, List, Optional, Type
+
+from requests import get
+from url_matcher import URLMatcher
+
+from . import default_registry
+from .page_inputs import HttpResponse, PageParams
+from .pages import ItemPage
+from .rules import consume_modules
+
+
+class _HttpClient:
+    async def get(self, url: str) -> HttpResponse:
+        return _get_http_response(url)
+
+
+def _get_page_class(url: str) -> Type[ItemPage]:
+    url_matcher = URLMatcher(
+        {rule.use: rule.for_patterns for rule in default_registry.get_rules()}
+    )
+    return url_matcher.match(url)
+
+
+def _get_http_response(url: str) -> HttpResponse:
+    response = get(url)
+    return HttpResponse(
+        response.url,
+        body=response.content,
+        headers=response.headers,
+    )
+
+
+def _get_page(
+    url: str,
+    page_class: Type[ItemPage],
+    *,
+    page_params: Optional[Dict[Any, Any]] = None,
+) -> ItemPage:
+    # We make quite a few assumptions here that prevent this function from
+    # working in scenarios where those assumptions are wrong. These shortcuts
+    # have been taken to save development time, providing the minimum
+    # implementation that works well enough for the tutorial.
```

Comment:
I think we should use andi even here, even if it's a bit more complicated. Hopefully, it wouldn't be too bad.

Reply:
https://github.com/scrapinghub/web-poet/pull/108/commits/94cf9a9886545ff20af204f99ad806aee13ddb12 switches to andi.

However, besides that, it is still rather tutorial-specific (only needed “providers” are implemented, and the HttpClient object only supports the interface subset used in the tutorial), and not too clean (e.g. we could organize providers in a way closer to scrapy-poet, and have a cleaner for look to build instances).

I’m not sure how far I should go with the implementation within the scope of these tutorial changes. Although the hard part, me getting to understand the basics of andi, is done now.